File: A/src/Checker.py
from modules.Cylinder import Cylinder
from modules.SmokeBall import SmokeBall
from modules.Missile import Missile

def cross(smokeBall, A, M):
    '''
    检测烟雾弹有没有让M找不到checkPoint
        
    Args:
        smokeBall: SmokeBall object 烟雾弹
        A: (x,y,z) tuple 目标点
        M: M object 导弹
    Returns:
        True or False 找没找到
    '''
    import math
    import numpy as np
    AM=(np.array([M.x,M.y,M.z])-np.array(A))
    AO=(np.array([smokeBall.x,smokeBall.y,smokeBall.z])-np.array(A))
    if np.dot(AO,AO)<=smokeBall.r**2:
        return True
    # if np.dot(AM,AO)<=0:
    #     return False
    cosOAM=np.dot(AM,AO)/(math.sqrt(np.dot(AM,AM))*math.sqrt(np.dot(AO,AO)))
    d=math.sqrt(np.dot(AO,AO))*math.sqrt(1-cosOAM**2)
    l=math.sqrt(np.dot(AO,AO))*cosOAM
    if d<=smokeBall.r and math.sqrt(np.dot(AM,AM))>=l-(smokeBall.r**2-d**2)**0.5:
        return True
    else:
        return False
    
def visible(checkCylinder,smokeBall,missile):
    '''
    检测导弹能不能看到真目标
    Args:
        checkCylinder: Cylinder object 真目标
        smokeBall: SmokeBall object 烟雾弹
        missile: Missile object 导弹
    Returns:
        True or False 能看到真目标 or 不能看到真目标
    '''
    import math
    import numpy as np
    OM=(np.array([missile.x,missile.y,missile.z])-np.array([smokeBall.x,smokeBall.y,smokeBall.z]))
    if np.dot(OM,OM)<=smokeBall.r**2:
        return False# 导弹在烟雾弹内，必然看不到真目标
    if missile.x<0:
        return True# 导弹已经炸了
    checkN=50
    for i in range(checkN):
        point=checkCylinder.random_point_on_surface()
        if not cross(smokeBall=smokeBall, A=point, M=missile):
            return True
    return False

File: A/rust/src/checker.rs
// Rust version of Checker.py
use crate::modules::{cylinder::Cylinder, smokeball::SmokeBall, missile::Missile};
use nalgebra::Vector3;

pub fn cross(A:(f64,f64,f64),O: &SmokeBall,M: &Missile)->bool{
    let AO=Vector3::new(O.x-A.0,O.y-A.1,O.z-A.2);
    let AM=Vector3::new(M.x-A.0,M.y-A.1,M.z-A.2);
    if AO.norm()<=O.r{
        return true;
    }
    let cosOAM=AO.dot(&AM)/(AO.norm()*AM.norm());
    let sinOAM=(1.0-cosOAM*cosOAM).sqrt();
    let d=AO.norm()*sinOAM;
    let l=AO.norm()*cosOAM;
    d<=O.r && AM.norm()>=l-(O.r*O.r-d*d).sqrt()
}

pub fn visible(check_cylinder: &Cylinder, smoke_ball: &SmokeBall, missile: &Missile) -> bool {
    let OM=Vector3::new(missile.x-smoke_ball.x,missile.y-smoke_ball.y,missile.z-smoke_ball.z);
    if OM.norm()<=smoke_ball.r{
        return false;
    }
    if missile.x<0.0{
        return true;
    }
    let checkN=500;
    for _i in 0..checkN{
        let point=check_cylinder.random_point_on_surface();
        if !cross(point,smoke_ball,missile){
            return true;
        }
    }
    false
}
